### ISSUE TITLE
Midi files are not playing

### ISSUE THREAD

**Issue by ntfwc** [ZQuestClassic/ZQuestClassic#121]
This affects both zelda and zquest. In zquest, if you try to play music from the Etc>The Travels of Link menu, nothing happens. Same result if you try to play a midi from a quest through Quest>Audio>MIDIs. In zelda, if you play a quest with midi music, the music does not play.

This problem does not occur in 2.50.2.

Investigation:
I was able to bisect the problem and found that the last good commit was 1ca34af68d3c009201e70373e70416dbb0c7260c, and the first known bad commit is f95b80061749ac4ead6fb1cbf76660b2e5540c6c (as the previous commit could not be built). Combing through the changes I discovered that this was introduced with the removal of two methods, file_size_ex_password and pack_fopen_password, from file.c and the introduction of alleg_compat.cpp. So there seems to be some problem with alleg_compat.cpp's re-implementation of these methods.

OS: Xubuntu 16.04 64-bit
commit: c8701fc9320a873e9542483c11398272ac0ccdad

**Comment by ZoriaRPG** [comment]
The problem is in the Allegro 4.4.2 / 4.4.3 library itself, and modern Linux systems refusing to load the Allegro MIDI driver. Thus far, none of us have found a solution, and the alegro.cc board community have not solved it, either.

**Comment by evouga** [comment]
According to ntfwc's investigation it sounds like the problem is not with the MIDI driver, but with how Allegro is loading MIDIs. Allegro changed around how packfile passwords are handled between 4.4.2 and 4.4.4 and it's possible that's the root issue here.

All of the Travels of Link MIDIs get loaded from zquest.dat in zquest.cpp line 20943 (in master). It's possible that the MIDI data is getting somehow corrupted there, though it's strange that it's only happening on Linux and that Allegro is not reporting an error when loading the datafile.

What do you get if you inspect (MIDI *)zcdata[9].dat after loading the datafile? I get 5 tracks with length 781, 63391, 134269, 59, and 70.

**Comment by ntfwc** [comment]
I do indeed get the same track lengths.

I have observed a separate midi failure, on 1ca34af68d3c009201e70373e70416dbb0c7260c, but only if my config file is missing the following:
```
[sound]
patches = samplesoundset/patches.dat
```
"The Travels of Link" music doesn't play, but the symptoms are different. "The Travels of Link" menu does get a checkmark next to it and I get an alsa midi device error on the command line. For this issue, it does not get a checkmark, and there is no command line output.

**Comment by evouga** [comment]
That one makes sense to me though: `patches` is a built-in Allegro variable that tells the Allegro sound engine where to find the MIDI patches.

**Comment by ntfwc** [comment]
I'm going to see if changing alleg_compat to provide identical methods as file.c in 1ca34af68d3c009201e70373e70416dbb0c7260c can restore the functionality on master.

**Comment by ntfwc** [comment]
Confirmed. A direct replication of the old methods does work. Now hopefully I can figure out a more elegant solution.

**Comment by ZoriaRPG** [comment]
This is what Gideon Weems said on the Allegro boards:

> You are either half right or half wrong; take your pick. A number of older programs I've compiled 
>against Allegro 4.4 on a fairly up-to-date system (kernel 4.12) have required MIDI_NONE in order for 
>install_audio() to successfully return. I recently compiled a SpeedHack entry from 2002, and the only 
>Allegro-specific patching it required was MIDI_NONE. (Other patching involved the C standard, which 
>has apparently changed more than Allegro's API over the same period of time.)
>
>As for possible explanations and solutions (for when MIDI is absolutely necessary), perhaps someone 
>closer to Allegro's source can comment.
>
>Anyway, what about the A4 example programs? Can you get MIDI output from them?

**Comment by ntfwc** [comment]
So I found a solution and, after a bunch of dprintf debugging, I was able to find the cause. Turns out it was fairly simple.

The problem occurs when it tries to load samplesoundset/patches.dat. The old methods left the global "the_password" field NULL, but the newer alleg_compat methods were leaving the last used password around. This changed the encrypt_id result when loading samplesoundset/patches.dat, resulting in it failing to load.

I will add a pull request.
